swarm-test/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+use async_trait::async_trait;
+use futures::future::Either;
+use futures::{AsyncRead, AsyncWrite, StreamExt};
+use libp2p::core::upgrade::Version;
+use libp2p::identity::Keypair;
+use libp2p::noise::NoiseAuthenticated;
+use libp2p::swarm::{
+    AddressScore, ConnectionHandler, IntoConnectionHandler, NetworkBehaviour, SwarmEvent,
+};
+use libp2p::yamux::YamuxConfig;
+use libp2p::{Multiaddr, PeerId, Swarm, Transport};
+use std::fmt::Debug;
+use std::time::Duration;
+
+type THandler<TBehaviour> = <TBehaviour as NetworkBehaviour>::ConnectionHandler;
+type THandlerErr<TBehaviour> =
+    <<THandler<TBehaviour> as IntoConnectionHandler>::Handler as ConnectionHandler>::Error;
+
+/// An extension trait for [`Swarm`] that makes it easier to set up a network of [`Swarm`]s for tests.
+#[async_trait]
+pub trait SwarmExt {
+    type NB: NetworkBehaviour;
+
+    /// Create a new [`Swarm`] with an ephemeral identity.
+    fn new_ephemeral<T>(
+        transport: T,
+        behaviour_fn: impl FnOnce(PeerId, Keypair) -> Self::NB,
+    ) -> Self
+    where
+        T: Transport + Send + 'static + Unpin,
+        T::Dial: Send + 'static,
+        T::ListenerUpgrade: Send + 'static,
+        T::Error: Send + Sync,
+        T::Output: AsyncRead + AsyncWrite + Send + 'static + Unpin,
+        Self: Sized;
+
+    /// Establishes a connection to the given [`Swarm`], polling both of them until the connection is established.
+    async fn block_on_connection<T>(&mut self, other: &mut Swarm<T>)
+    where
+        T: NetworkBehaviour + Send,
+        <T as NetworkBehaviour>::OutEvent: Debug;
+
+    /// Listens on a random memory address, polling the [`Swarm`] until the transport is ready to accept connections.
+    async fn listen_on_random_memory_address(&mut self) -> Multiaddr;
+
+    /// Listens on a random memory address, polling the [`Swarm`] until the transport is ready to accept connections.
+    async fn listen_on_random_localhost_tcp_port(&mut self) -> Multiaddr;
+
+    async fn next_within(
+        &mut self,
+        seconds: u64,
```

Comment:
Having to provide a timeout seems off to me. How would a user decide on the right value? Can't we just rely on the overall test timeout?

Reply:
Tests don't have an overall timeout unless the user wraps it all in one which is boilerplate and thus doesn't happen very often. Plus, compared to a test timeout, having the test failure in here provides a nice stacktrace that is easier to debug on where the test is failing.

The reason we can't have a timeout default is because we don't know about the test case, only the dev writing the test can decide that :)

Reply:
> Tests don't have an overall timeout unless the user wraps it all in one which is boilerplate and thus doesn't happen very often. Plus, compared to a test timeout, having the test failure in here provides a nice stacktrace that is easier to debug on where the test is failing.

I wasn't aware of this. Thanks.

Still, I would consider concrete timeouts in tests a code smell. Users of this framework would have to make assumptions on the environment this is run on. E.g. they need to take into account that their test might run on a GitHub CI or some high performant developer machine.

I can not think of a way without the timeout while still having a strack trace. Is the strack trace important? I guess this boils down to "should one be able to see the root cause purely from the unit test failure, or do we expect a red test to need debugging anyways".

I don't expect any unit test to take longer than 10 seconds. In case we want to keep the stack traces, how about setting the timeout to a fixed `10s`?

Reply:
> I don't expect any unit test to take longer than 10 seconds. In case we want to keep the stack traces, how about setting the timeout to a fixed `10s`?

I guess for any situation where the timeout doesn't fit, the user can always fall back to the `StreamExt::next` function which doesn't have a timeout so this is probably a good idea?

What we need to balance with these test abstractions is for them to flexible enough to handle a lot of situations but also providing enough value in making test code concise. I think a timeout 10s is a good default and if it doesn't fit, you can always just code it yourself and it is not too verbose. If it happens a lot, we can look into adjusting the timeout or finding other solutions.